### ISSUE TITLE
[T10] Visual encoder: window pixels and position to input currents

### ISSUE THREAD

**Issue by akhmialeuski** [akhmialeuski/fly-connectome-lab#11]
## Goal

A frozen, deterministic encoder that turns one RGB window and its position into per-neuron currents for a fixed set of MaleCNS visual projection neurons. The encoder is intentionally simple so that the recognition work is done by the connectome, not by the encoder.

## Depends on

- T04, T08

## Files to create

```text
src/flystate/encoders/__init__.py
src/flystate/encoders/sparse_projection.py
tests/unit/test_encoder.py
```

## API

```python
class SparseProjectionEncoder:
    def __init__(self, cfg: EncoderConfig, window: int, candidate_neurons: np.ndarray) -> None: ...
    input_idx: np.ndarray        # (M,) int64 neuron indices, order: R group, G group, B group, pos-x, pos-y
    def encode(self, observations: np.ndarray, positions: np.ndarray) -> np.ndarray: ...
    def spec(self) -> dict[str, object]: ...
```

- `candidate_neurons` are the indices returned by `EpisodeBrain.cells([cfg.target_population])`.
- `encode(observations (B, W, W, 3) uint8, positions (B, 2) float32) -> currents (M, B) float32`, ready for `EpisodeBrain.run(input_idx, currents, k)`.

## Construction, exactly

- `P = cfg.neurons_per_channel`, `Q = cfg.position.neurons_per_axis` when enabled, else 0. `M = 3 * P + 2 * Q`. Raise `ValueError` when `M > len(candidate_neurons)` (the real brain has 9,201 visual projection neurons; the default needs 3,872).
- Neuron choice: `rng = Generator(PCG64(SeedSequence([cfg.seed, stable_int("encoder-neurons")])))`; `chosen = rng.permutation(np.sort(candidate_neurons))[:M]`; groups are consecutive slices of `chosen` in the order R, G, B, pos-x, pos-y.
- Projection for channel `c` in (0, 1, 2): `rng_c = Generator(PCG64(SeedSequence([cfg.seed, stable_int(f"encoder-projection-{c}")])))`; a `scipy.sparse.csr_matrix` `M_c` of shape `(P, W*W)`, float32, where row `i` has exactly `cfg.nonzeros_per_row` non-zeros at columns `rng_c.choice(W*W, nonzeros_per_row, replace=False)` with values `rng_c.choice([-1, 1]) / sqrt(nonzeros_per_row)` each.

## Encoding, exactly

- `x_c = 2 * observations[..., c].reshape(B, W*W) / 255 - 1` (float32, in [-1, 1]).
- Pixel groups: `u_c = a * clip(M_c @ x_c.T, -1, 1)` with shape `(P, B)`, `a = cfg.amplitude`.
- Position groups: centres `c_j = (j + 0.5) / Q`, `sigma = 1 / Q`; `u_x[j, b] = a * exp(-(pos[b, 0] - c_j)**2 / (2 * sigma**2))`, the same for `y` with `pos[b, 1]`.
- Result: vertical stack `[u_R; u_G; u_B; u_x; u_y]`, float32, shape `(M, B)`.

`spec()` returns `{"kind", "window", "neurons_per_channel", "nonzeros_per_row", "position", "amplitude", "seed", "input_idx_sha256", "projection_sha256"}`; the hashes are sha256 of the raw bytes of `input_idx` and of the concatenated `indices`, `indptr`, `data` arrays of the three matrices.

## Tests (synthetic brain with 600 visual projection neurons; use `neurons_per_channel = 64`, `neurons_per_axis = 8`)

- Determinism: two encoders with the same config have identical `input_idx`, identical `spec()` and identical output.
- Every projection row has exactly `nonzeros_per_row` non-zeros; values are `±1/sqrt(nonzeros_per_row)`.
- Groups are disjoint and all indices belong to `candidate_neurons`.
- `|currents| <= a` everywhere; a black window (all 0) and a white window (all 255) give opposite pixel currents before clipping (check on an unclipped case with one non-zero per row).
- For position `(0.5, 0.5)` the maximum of the x group is at `j = Q/2 - 1` or `Q/2`.
- `M > len(candidate_neurons)` raises `ValueError`.
- A different `seed` changes `input_idx`.

## Acceptance criteria

Definition of done from `AGENTS.md`.

## Out of scope

Retinotopic encoders through photoreceptors (`brain.npz` has only a 1-D azimuth per photoreceptor, and photoreceptor input does not pass the lamina in this spiking model).

**Comment by akhmialeuski** [comment]
## Recorded encoder and outstanding diagnostic questions

Evidence backfill dated 2026-09-21. Research follow-up: #39.

The frozen encoder uses fixed seeded sparse signed RGB projections into selected visual-projection neurons, with separate deterministic position channels. It is not a pretrained visual model. The effective configuration and calibration records preserve input-selection/projection identities.

Observed evidence does not yet localize the recognition failure to this encoder: pixel controls retain identity information, while the final neural readout is near chance, but intermediate encoded-input classification has not yet been measured. The lowest amplitude producing a detectable downstream response is not necessarily the amplitude preserving the most identity information.

Plan #39 therefore compares the same simple probe before and after encoding, then studies noise/duration and anatomical input mapping if warranted. No specialized model, learned face embedding, or hybrid pretrained encoder is part of that plan. A two-dimensional retinal map must not be invented from the currently available azimuth-only input metadata.

Source: [encoder](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/src/flystate/encoders/sparse_projection.py). Effective configurations: #3. Calibration reports: #12.

[Study README and restoration instructions](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/README.md) · [Complete SHA-256 inventory](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/checksums.sha256). Large payloads require Git LFS; these links are pinned to the archived commit.

**Comment by akhmialeuski** [comment]
## Completed numerical follow-up and valid input controls

Scientific execution commit: `f821470ac263a584eb2a42e05108030d7fbb2630`. The numerical protocol was registered in [comment 5763354124](https://github.com/akhmialeuski/fly-connectome-lab/issues/40#issuecomment-5763354124); full retries were registered before scoring in [comment 5769606778](https://github.com/akhmialeuski/fly-connectome-lab/issues/40#issuecomment-5769606778).

All four original failures reproduce as iteration-limit stops at 5,000. Increasing only the iteration cap converges on the same first failed fold and C in 5,411–10,396 iterations; every final gradient infinity norm is below 1e-6. Independent replay of all eight exported fold parameter sets reproduces objective values within 1.12e-16 and gradient norms within 1.93e-15. These fold models are numerical evidence, not selected recognition models.

All four complete five-fold C-grid retries then converge under the common 50,000-iteration cap, unchanged tolerance 1e-6, original fold-local scaler/PCA60, and seed 0. There is no discarded C or relaxed accuracy criterion.

| Fresh attempt | Training correct | Validation correct | Validation top-1 | Validation log loss | Selected C | Best training CV |
| --- | --- | --- | --- | --- | --- | --- |
| retry50000-smoke-02-pixels-persistent-all-both-pca60-true | 280/280 | 31/60 | 51.67% | 3.552894 | 1.0 | 47.14% |
| retry50000-smoke-04-encoded-persistent-all-both-pca60-true | 280/280 | 31/60 | 51.67% | 2.299052 | 0.01 | 44.29% |
| retry50000-main-01-pixels-persistent-all-both-pca60-true | 1400/1400 | 112/300 | 37.33% | 3.960701 | 0.01 | 36.36% |
| retry50000-main-02-encoded-persistent-all-both-pca60-true | 1400/1400 | 107/300 | 35.67% | 4.312019 | 0.01 | 33.93% |

### Interpretation

- Identity information is measurably available in the original pixels and in the fixed encoded currents. Smoke: both full-history controls score 31/60 (51.67%), versus the frozen persistent-last neural anchor at 4/60 (6.67%) and 5% chance. Main: pixels score 112/300 (37.33%) and encoded currents 107/300 (35.67%), versus neural persistent-last at 1/300 (0.33%), neural reset-all at 2/300 (0.67%), and 1% chance.
- The cached neural PCA/feature/history screening did not pass its preregistered promotion gate. This localizes the practical failure to the measured neural representation/readout portion of the current pipeline, rather than complete absence of information in the input. It does not identify a specific biological layer, prove the encoder optimal, or distinguish noise, population access, timing, quantization, and state/readout filtering as causes.
- The original optimizer failures were numerical failures, not zero recognition. The historical pixel scores used a different tolerance (1e-4); retain them as historical measurements, not as replacements for these matched 1e-6 controls. Numerical stopping can affect CV selection and held-out scores, so report the actual repeated controls.
- Training accuracy alone remains inadequate evidence: every new input control fits training perfectly; the smoke neural model also memorizes training without established useful generalization. No useful recognition or sequential-memory advantage has been established for the fly model.
- Every score above is exploratory validation. No old final-test or reserved-photo predictions were scored. The independent confirmation limitation from the cohort audit remains.

### Next decisions

Proceed with the fixed-cohort learning curve in #42 to address the explicit training-size question. Its 48 unique attempts and memberships are frozen before fitting. Both cohorts previously had only 14 training photos per identity. Then continue the separate same-image/noise-seed, direct float32/float16, and signal-delivery diagnostics required by #39. Do not jump from this readout result to arbitrary equation changes or claim that additional data or constrained plasticity cannot help.

### Evidence and engineering checks

The dated diagnostic study retains 26 full-probe attempts (22 completed models and four original failures), two cohort audits, and four numerical diagnoses with eight exported fold parameter sets. All per-attempt reports were posted to #40. Model coefficients, predictions, failures, provenance, and checksum inventories are being appended to `research/sequential-visual-memory/2026-09-21-fly-only-diagnostics/`; remote verification is reported separately after publication.

The numerical implementation passed 313 tests with two optional skips, static checks, pre-commit, and source/wheel builds; its CI check passed. During the independent verification script, an incorrect two-value unpack of the three-value objective/gradient API caused a verifier-only exception. The local script was corrected to combine coefficient and intercept gradient norms, and all eight fold models then replayed successfully. Both verifier logs are retained. This did not modify or rerun a scientific fit.
